core/extension/gdnative_interface.h

```diff
@@ -0,0 +1,418 @@
+	void (*string_new_with_wide_chars_and_len)(GDNativeStringPtr r_dest, const wchar_t *p_contents, const GDNativeInt p_size);
+	const char *(*string_to_latin1_chars)(const GDNativeStringPtr p_self);
+	const char *(*string_to_utf8_chars)(const GDNativeStringPtr p_self);
+	const char16_t *(*string_to_utf16_chars)(const GDNativeStringPtr p_self);
+	const char32_t *(*string_to_utf32_chars)(const GDNativeStringPtr p_self);
+	const wchar_t *(*string_to_wide_chars)(const GDNativeStringPtr p_self);
+	char32_t *(*string_operator_index)(GDNativeStringPtr p_self, GDNativeInt p_index);
+	const char32_t *(*string_operator_index_const)(const GDNativeStringPtr p_self, GDNativeInt p_index);
+
+	/* OBJECT */
+
+	void (*object_method_bind_ptrcall)(GDNativeMethodBindPtr p_method_bind, GDNativeObjectPtr p_instance, const GDNativeTypePtr *p_args, GDNativeTypePtr r_ret);
+	void (*object_destroy)(GDNativeObjectPtr p_o);
+	GDNativeObjectPtr (*global_get_singleton)(const char *p_name);
+
+	GDNativeObjectPtr (*object_cast_to)(const GDNativeObjectPtr p_object, void *p_class_tag);
+	GDNativeObjectPtr (*object_get_instance_from_id)(GDObjectInstanceID p_instance_id);
+	GDObjectInstanceID (*object_get_instance_id)(const GDNativeObjectPtr p_object);
+
+	/* CLASSDB */
+
+	GDNativeClassConstructor (*classdb_get_constructor)(const char *p_classname);
+	GDNativeMethodBindPtr (*classdb_get_method_bind)(const char *p_classname, const char *p_methodname, GDNativeInt p_hash);
+	void *(*classdb_get_class_tag)(const char *p_classname);
+
+	/* CLASSDB EXTENSION */
+
+	void (*classdb_register_extension_class)(const GDNativeExtensionClassLibraryPtr p_library, const char *p_class_name, const char *p_parent_class_name, const GDNativeExtensionClassCreationInfo *p_extension_funcs);
+	void (*classdb_register_extension_class_method)(const GDNativeExtensionClassLibraryPtr p_library, const char *p_class_name, const GDNativeExtensionClassMethodInfo *p_method_info);
+	void (*classdb_register_extension_class_integer_constant)(const GDNativeExtensionClassLibraryPtr p_library, const char *p_enum_name, const char *p_class_name, const char *p_constant_name, uint32_t p_constant_value);
+	void (*classdb_register_extension_class_property)(const GDNativeExtensionClassLibraryPtr p_library, const char *p_class_name, const GDNativePropertyInfo *p_info, const char *p_setter, const char *p_getter);
+	void (*classdb_register_extension_class_signal)(const GDNativeExtensionClassLibraryPtr p_library, const char *p_class_name, const char *p_signal_name, const GDNativePropertyInfo *p_argument_info, GDNativeInt p_argument_count);
+} GDNativeInterface;
+
+/* INITIALIZATION */
+
+typedef enum {
+	GDNATIVE_INITIALIZATION_CORE,
+	GDNATIVE_INITIALIZATION_SERVERS,
+	GDNATIVE_INITIALIZATION_SCENE,
+	GDNATIVE_INITIALIZATION_EDITOR,
+} GDNativeInitializationLevel;
+
+typedef struct {
+	/* Minimum initialization level required.
+	 * If Core or Scene, the extension needs editor or game restart to take effect */
```

Comment:
if core&scene levels requires restart, wouldn't it be more logical to change the order of the initialization level:
```
typedef enum {
	GDNATIVE_INITIALIZATION_CORE,
	GDNATIVE_INITIALIZATION_SCENE,
        # From this on, the extension can be restarted without game/editor restart
	GDNATIVE_INITIALIZATION_SERVERS,
	GDNATIVE_INITIALIZATION_EDITOR,
} GDNativeInitializationLevel;
```

Though it seems odd to me that scene is loaded earlier than servers, so I guess `GDNativeInitializationLevel` order is right (i.e. it reflect the actual initialization order), but scene has some special stuff that prevent it from being able to hot restart.

In any way I think more comment would be welcome here (especially given this file is aimed at being read by extension developer with lesser to no understanding of Godot internal)

Reply:
It might be also good to explicitly number `GDNativeInitializationLevel` values to underline the initialization order:
```c
typedef enum {
	GDNATIVE_INITIALIZATION_CORE = 0,
	GDNATIVE_INITIALIZATION_SERVERS = 1,
	GDNATIVE_INITIALIZATION_SCENE = 2,
	GDNATIVE_INITIALIZATION_EDITOR = 3,
} GDNativeInitializationLevel;
```
or maybe at least to add a comment about the loading order
```c
typedef enum {
        // Initialization order, from first to last
	GDNATIVE_INITIALIZATION_CORE,
	GDNATIVE_INITIALIZATION_SERVERS,
	GDNATIVE_INITIALIZATION_SCENE,
	GDNATIVE_INITIALIZATION_EDITOR,
} GDNativeInitializationLevel;
```

Reply:
my mistake there again, I meant core and servers require restart